lib/telemetry_metrics.ex

```diff
@@ -1,5 +1,341 @@
+    which are set as `:tags`. There are three possible values:
+    * `:all` - all event metadata is used;
+    * list of terms, e.g. `[:table, :kind]` - only these keys from the event metadata are used;
+    * one argument function taking the event metadata and returning the metadata which should be
+      used to generate tag values
+  * `:description` - human-readable description of the metric. Might be used by reporters for
+    documentation purposes. Defaults to `nil`;
+  * `:unit` - the unit of event values. Might be used by reporters for documentation purposes.
+    Defaults to `:unit`.
+
+  ## Reporters
+
+  Reporters take metric definitions as an input, subscribe to relevant events and update the metrics
+  when the events are emitted. Updating the metric might involve publishing the metrics periodically,
+  or on demand, to external systems. `Telemetry.Metrics` defines only specification for metric types,
+  and reporters should provide actual implementation for these aggregations.
+
+  ### Rationale
+
+  The design proposed by `Telemetry.Metrics` might look controversial - unlike most of the libraries
+  available on the BEAM, it doesn't aggregate metrics itself, it merely defines what users should
+  expect when using the reporters. There are two arguments for this solution.
+  if `Telemetry.Metrics` would aggregate metrics, the way those aggregations work would be imposed
+  on the system where the metrics are published to. For example, counters in StatsD are reset on
+  every flush and can be decremented, whereas counters in Prometheus are monotonically increasing.
+  `Telemetry.Metrics` doesn't focus on those details - instead, it describes what the end user,
+  operator, expects to see when using the metric of particular type. This implies that in most
+  cases aggregated metrics won't be visible inside the BEAM, but in exchange aggregations can be
+  implemented in a way that makes most sense for particular system. Finally, one could also
+  implement an in-VM "reporter" which would aggregate the metrics and expose them inside the BEAM.
+  When there is a need to swap the reporters, and if both reporters are following the metric types
+  specification, then the end result of aggregation is the same, regardless of the backend system
+  in use.
+
+  ### Requirements for reporters
+
+  Reporters should accept metric specifications and subscribe to relevant events. When those events
+  are emitted, metric should be updated (either in-memory or by contacting external system) in such
+  a way that the user is able to view metric values as described in the "Metric types" section.
+
+  If the reporter does not support the metric given to it, it should log a warning.
+
+  Reporters should also document how `Telemetry.Metrics` metric types, names tags are translated to
+  metric types and identifiers in the system they publish metrics to.
   """
+
```

Comment:
Do not forget to define a `@type t :: %{__struct__: module, name: metric_name, ...}` with all of the common fields. :)

Reply:
It's in `lib/telemetry_metrics/metric.ex`. Do you think it should be placed in the top-level module?

Reply:
Yeah, I don't think we need a module only for that, so I would definitely move it up here. :)